src/cbrkit/sim/graphs/_alignment.py

```diff
@@ -0,0 +1,174 @@
+from __future__ import annotations
+
+from dataclasses import dataclass
+from typing import Callable, List, Tuple, Any
+from ._model import Graph, Node, is_sequential
+from ...helpers import dist2sim
+
+__all__ = []
+
+try:
+    import numpy as np
+    from ..collections import dtw
+
+    @dataclass(slots=True)
+    class GraphDTW:
+        """
+        Graph-based Dynamic Time Warping similarity function leveraging sequence alignment.
+
+        Examples:
+            >>> # Example node and edge similarity functions
+            >>> node_similarity = lambda n1, n2: 1.0 if n1.data == n2.data else 0.0
+            >>> edge_similarity = lambda e1, e2: 1.0 if e1.data == e2.data else 0.0
+            >>> # Create example graphs
+            >>> from src.cbrkit.sim.graphs._model import Graph, Node, Edge
+            >>> import immutables
+            >>> nodes1 = immutables.Map({
+            ...     '1': Node(key='1', data='A'),
+            ...     '2': Node(key='2', data='B'),
+            ... })
+            >>> edges1 = immutables.Map({
+            ...     'e1': Edge(key='e1', source=nodes1['1'], target=nodes1['2'], data='X'),
+            ... })
+            >>> graph1 = Graph(nodes=nodes1, edges=edges1, data=None)
+            >>> nodes2 = immutables.Map({
+            ...     '1': Node(key='1', data='A'),
+            ...     '2': Node(key='2', data='C'),
+            ... })
+            >>> edges2 = immutables.Map({
+            ...     'e1': Edge(key='e1', source=nodes2['1'], target=nodes2['2'], data='Y'),
+            ... })
+            >>> graph2 = Graph(nodes=nodes2, edges=edges2, data=None)
+            >>> # Instantiate GraphDTW
+            >>> g_dtw = GraphDTW(node_similarity, edge_similarity)
+            >>> result = g_dtw(graph1, graph2)
+            >>> print(result)
+            (0.3333333333333333, [('1', '1'), ('2', '2')])
+        """
+
+        node_sim_func: Callable[[Any, Any], float]
+        edge_sim_func: Callable[[Any, Any], float] | None = None  # Optional for edge similarity
+        normalize: bool = True  # Whether to normalize similarity by sequence length
+
+        def __call__(
+                self,
+                graph1: Graph[K, N, E, G],
+                graph2: Graph[K, N, E, G]
+        ) -> Tuple[float, List[Tuple[K, K]]]:
+            """
+            Perform Graph-DTW using node and optional edge similarity functions.
+
+            Args:
+                graph1: The first graph.
+                graph2: The second graph.
```

Comment:
Since a mapping is constructed, it should be clear which one is which. In CBRkit, the first argument of a similarity function is always x (the case) and the second y (the query). The mapping shall be constructed from y to x